---
title: 'FC'
author: 'ABO'
date: 'created : 2025-11-04 — compiled : `r format(Sys.time(), "%Y-%m-%d %H:%M")`'
output:
  html_document:
    code_folding: hide
    toc: true # Table des matières
    # toc_float: 
      # collapsed: false # Commencer avec le TOC déplié
      # smooth_scroll: true # Permet un défilement fluide jusqu'à la section sélectionnée
    # toc_depth: 3 # Profondeur du TOC (niveaux de titres inclus)
    number_sections: true # Numérotation des sections
    css: style.css # Fichier CSS personnalisé
    # df_print: paged # Mode d'affichage des data frames (options : default, paged, kable, tibble, none)
  pdf_document:
    toc: true # Table des matières
    toc_depth: 3 # Profondeur du TOC
    # number_sections: true # Numérotation des sections
    # latex_engine: xelatex # Moteur LaTeX (pdflatex, xelatex, lualatex)
    citation_package: natbib # Système de gestion des citations (natbib, biblatex)
header-includes:
  # - \usepackage{lipsum} # Inclure des packages LaTeX supplémentaires ici
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
# CORE
library(tidyverse)
library(janitor)
library(readxl)
library(lubridate)
# FOR REPORT
library(broom)
library(plotly)
library(patchwork)
library(kableExtra)
library(randomcoloR)
library(xml2)
```

```{r FUNCTIONS}
extract_variant_info <- function(xml_file_path) {
  cat(xml_file_path, file=stderr())
  
  # Charger le fichier XML
  xml_file <- read_xml(xml_file_path)
  
  # Obtenir le noeud racine
  root_node <- xml_root(xml_file)
  
  # Trouver tous les noeuds 'variant'
  variant_nodes <- xml_find_all(root_node, ".//variant")
  
  # Extraire le nom du fichier XML sans son chemin complet
  xml_file_name <- basename(xml_file_path)
  
  # Mapper sur chaque 'variant' pour extraire les informations
  variants_data <- map_df(variant_nodes, function(variant) {
    
    # Extraire les noeuds 'refseq'
    refseq_nodes <- xml_find_all(variant, ".//refseq")
    
    # Mapper sur chaque 'refseq' pour obtenir les détails
    map_df(refseq_nodes, function(refseq) {
      tibble(
        chromosome = xml_text(xml_find_first(variant, ".//chromosome")),
        position = xml_text(xml_find_first(variant, ".//position")),
        reference = xml_text(xml_find_first(variant, ".//reference")),
        alternative = xml_text(xml_find_first(variant, ".//alternative")),
        gene = xml_text(xml_find_first(refseq, ".//gene")),
        exon = xml_text(xml_find_first(refseq, ".//exon")),
        function_type = xml_text(xml_find_first(refseq, ".//function")),
        transcript = xml_text(xml_find_first(refseq, ".//transcript")),
        nucleotidic_position = xml_text(xml_find_first(refseq, ".//nucleotidic_position")),
        proteinic_position = xml_text(xml_find_first(refseq, ".//proteinic_position")),
        prefered_transcript = xml_text(xml_find_first(refseq, ".//prefered_transcript")),
        xml_file_name = xml_file_name  # Ajouter le nom du fichier XML
      )
    })
  })


  variants_data <- variants_data # %>% filter(prefered_transcript == "YES")
  
  return(variants_data)
}

# Fonction pour obtenir la liste des fichiers XML dans un dossier
get_xml_files <- function(folder_path) {
  list.files(path = folder_path, pattern = "\\.xml$", full.names = TRUE)
}

# Appliquer la fonction extract_variant_info à tous les fichiers XML dans le dossier
process_xml_folder <- function(folder_path) {
  xml_files <- get_xml_files(folder_path)
  
  # Utiliser map_df pour appliquer la fonction à chaque fichier et combiner les résultats en un seul DataFrame
  all_variants_data <- map_df(xml_files, extract_variant_info)
  
  return(all_variants_data)
}

```

# BIG
```{r}
################################################################################
# BIG DATA 
################################################################################

(
  read_xlsx("./input/bm/thèque Big1.xlsx")
  %>% clean_names()
  %>% mutate(type = str_to_upper(type))
) -> raw_bm_BIG

(
  raw_bm_BIG
  %>% group_by(n_inclusion)
  %>% fill(initiale_nom:stade, .direction = "down")
  %>% mutate(n_bm = ifelse(n_inclusion == "12A-010" & stade == "J60 post allo",
  "L1710153", n_bm))
  %>% mutate(n_bm = ifelse(n_bm == "L1714848", "L1714853", n_bm))
  %>% ungroup()
  %>% add_row(n_inclusion = "12A-020", n_bm = "L1816858")
) -> bm_BIG

(
  read_xlsx("./input/bm/BASE BIG1 FINALE 2023 patients 22_07_2024_IDENTITES_POSMYE.xlsx", sheet = 2)
  %>% clean_names()
) -> raw_POSMYE
```

# 101125
```{r 101125}
################################################################################
# SGE FILE 101125
################################################################################

(
  read_xls("./input/bm/TW-MRD-101125-DRAGEN.xls")
  %>% clean_names()
  %>% filter(projet == "MAGNET_AML")
  %>% count(patient)
) -> nom_run

(
  read_xls("./input/bm/TW-MRD-101125-DRAGEN.xls")
  %>% clean_names()
  %>% filter(projet == "MAGNET_AML")
  %>% select(prelevement)
  %>% mutate(to_keep = TRUE)
  %>% mutate(n_bm = str_remove(prelevement, "D[0-9]{2}"))
) -> ndosbm_run

(
  bm_BIG
  %>% left_join(ndosbm_run)
  %>% group_by(n_inclusion)
  %>% filter(any(to_keep == TRUE))
  %>% filter(to_keep == TRUE | stade == "diag")
  %>% arrange(n_inclusion, desc(stade), desc(ngs))
  %>% group_by(n_inclusion, stade)
  %>% slice(1)
  %>% arrange(n_inclusion, desc(stade), desc(ngs))
  %>% group_by(n_inclusion)
  %>% mutate(n = n())
) -> joined

# (
#   joined 
#   %>% filter(stade %in% c("J30 post allo", "J100 post allo", "3mois post allo", "post allo"))
#   %>% select(n_inclusion, nom, prenom, stade, n_bm)
#   %>% write_tsv(., pipe("pbcopy"))
# ) -> NDU_postG

# (
#   joined
#   # %>% filter(stade == "diag")
# ) -> diag

# (
#   joined
#   %>% mutate(to_keep = ifelse(!is.na(ngs), TRUE, to_keep))
# ) -> perfect_patient


################################################################################
# POSMYE
################################################################################

(
  read_xlsx("./input/bm/BASE BIG1 FINALE 2023 patients 22_07_2024_IDENTITES_POSMYE.xlsx", sheet = 2)
  %>% clean_names()
) -> raw_POSMYE

(
  bm_BIG
  %>% left_join(ndosbm_run)
  %>% filter(to_keep == TRUE)
) -> n_incl_SGE

(
  raw_POSMYE
  %>% filter(id %in% n_incl_SGE$n_inclusion)
) -> POSMYE_101125

(
  POSMYE_101125
  %>% filter(!is.na(x7))
  %>% mutate(x6 = str_extract(x6, "L.*"))
  # %>% select(x6)
) -> SQXT_ndosbm

# write_tsv(SQXT_ndosbm, "./input/bm/XML/SSQXT_101125.tsv")

################################################################################
# OPEN XML
################################################################################

all_variants_data <- process_xml_folder("./input/bm/XML/selected_101125/")

(
  all_variants_data
  %>% mutate(ndosbm = str_remove(xml_file_name, "D[0-9]{2}.xml"))
  %>% select(ndosbm, nucleotidic_position, chromosome, position, reference, alternative)
  %>% distinct()
  %>% ungroup()
) -> variant_data

list_gene_LOVER_2023 <- "ABCB11|ABL1|ABO|ANKRD26|ARID1A|ASXL1|ASXL2|ATG2B|ATM|ATRX|B2M|BAX|BCL2|BCOR|BCORL1|BIRC3|BRAF|BTG1|BTK|CALR|CARD11|CBL|CCL22|CCND1|CCND3|CCR4|CD274|CD28|CD58|CD79A|CD79B|CDKN2A|CDKN2B|CEBPA|CHEK2|CIITA|COG1|CREBBP|CRLF2|CSF2RA|CSF2RB|CSF3R|CTCF|CUX1|CXCR4|DDX3X|DDX41|DHX15|DHX34|DNMT3A|EBF1|EGLN1|EGR2|EP300|EPAS1|EPO|EPOR|ERG|ETNK1|ETV6|EVC|EZH2|F8|F9|F9|FBXW7|FERMT1|FGFR1|FIP1L1|FLT3|FOXO1|GATA1|GATA2|GNA13|GNAS|GNB1|GRM1|GSKIP|HRAS|ID3|IDH1|IDH2|IKZF1|IKZF3|IL2RG|IL3RA|IL5RA|IL7R|IRF4|ITPKB|JAK1|JAK2|JAK3|JUNB|KDM6A|KIT|KLF2|KMT2A|KMT2D|KRAS|L2HGDH|LAMA3|LUC7L2|MAP2K1|MAP3K14|MBD4|MEF2B|MPL|MSC|MYC|MYD88|NDUFV3|NF1|NFE2|NIPBL|NOTCH1|NOTCH2|NPHS2|NPM1|NRAS|NSD2|WHSC1|P2RY8|PAX5|PDGFRA|PDGFRB|PHF6|PIGA|PIM1|PLCG1|PLCG2|POT1|PPM1D|PRDM1|PROC|PROS1|PRPF8|PTEN|PTPN11|PTPRD|RAD21|RB1|RHOA|RIT1|RPS15|RUNX1|SAMD9|SAMD9L|SBDS|SERPINC1|SETBP1|SETD2|SF3B1|SH2B3|SLC12A6|SMC1A|SMC3|SOCS1|SOX6|SPI1|SRP72|SRSF2|SRY|STAG2|STAT1|STAT3|STAT5A|STAT5B|STAT6|TBL1XR1|TCF3|TCL1A|TDRD7|TERC|TERT|TET2|TNFAIP3|TNFRSF14|TP53|TYK2|U2AF1|UBA1|UBTF|VAV1|VCAN|VHL|VWF|WNK1|WT1|XPO1|ZBTB7A|ZRSR2"

################################################################################
# 54 p avec run SSQXT (2 NEG)
################################################################################

(
  SQXT_ndosbm
  %>% separate_rows(posmye, sep = "\r\n")
  %>% mutate(gene = str_extract(posmye, list_gene_LOVER_2023))
  %>% mutate(vaf = str_extract(posmye, "\\d+(\\.\\d+)?%")) 
  %>% mutate(vaf = as.numeric(str_remove(vaf, "%")))
  %>% mutate(c. = str_extract(posmye, "c.* :"))
  %>% mutate(c. = ifelse(is.na(c.), str_extract(posmye, "c.* \\("),c.))
  %>% mutate(c. = str_remove(c., "( :)| \\("))
  %>% mutate(p. = str_extract(posmye, "p\\..*\\("))
  %>% mutate(p. = str_remove(p., "\\("))
  %>% mutate(p. = ifelse(is.na(p.), "p.intron", p.))
  %>% mutate(mutation = paste0(gene, "_", p.))
  %>% mutate(gene = case_when(
    gene == "FLT3" & str_detect(c., "dup|in") ~ "FLT3-ITD",
    gene == "FLT3" & str_detect(c., ">") ~ "FLT3-TKD",
    TRUE ~ gene
  ))
  %>% select(-p.)
  %>% left_join(variant_data, by=c("ndosbm", "c." = "nucleotidic_position"))
  %>% group_by(gene, `c.`)
  %>% fill(chromosome, position, reference, alternative,
       .direction = "downup")
  %>% ungroup()
  %>% filter(posmye != "Analyse négative")
) -> SQXT_ndosbm_row

length(unique(SQXT_ndosbm_row$ndosbm))

(
  POSMYE_101125
  %>% filter(is.na(x7))
  %>% mutate(x6 = str_extract(x6, "L.*"))
  # %>% select(ndosbm)
  # %>% write_tsv(., pipe("pbcopy"))
  %>% filter(posmye != "Analyse négative.")
  %>% filter(posmye != ".")
) -> ndosbm_missing

(
  read_tsv("./input/bm/XML/selected_missing_101125/missing_ndosbm_101125.tsv",
  col_names = c("name", "path", "date"))
  %>% filter(!str_detect(path, "TW-MRD|BIOCOMPUTING|GROUP"))
  %>% arrange(name, desc(date))
  %>% group_by(name)
  %>% slice(1)
  %>% ungroup()
) -> path_ndosbm_missing

all_variants_data <- process_xml_folder("./input/bm/XML/selected_missing_101125/")

(
  all_variants_data
  %>% mutate(ndosbm = str_remove(xml_file_name, "D[0-9]{2}.xml"))
  %>% select(ndosbm, nucleotidic_position, chromosome, position, reference, alternative)
  %>% distinct()
  %>% ungroup()
) -> variant_data


(
  ndosbm_missing
  %>% separate_rows(posmye, sep = "\r\n")
  %>% mutate(gene = str_extract(posmye, list_gene_LOVER_2023))
  %>% mutate(vaf = str_extract(posmye, "\\d+(\\.\\d+)?%")) 
  %>% mutate(vaf = as.numeric(str_remove(vaf, "%")))
  %>% mutate(c. = str_extract(posmye, "c.* :"))
  %>% mutate(c. = ifelse(is.na(c.), str_extract(posmye, "c.* \\("),c.))
  %>% mutate(c. = str_remove(c., "( :)| \\("))
  %>% mutate(p. = str_extract(posmye, "p\\..*\\("))
  %>% mutate(p. = str_remove(p., "\\("))
  %>% mutate(p. = ifelse(is.na(p.), "p.intron", p.))
  %>% mutate(mutation = paste0(gene, "_", p.))
  %>% mutate(gene = case_when(
    gene == "FLT3" & str_detect(c., "dup|in") ~ "FLT3-ITD",
    gene == "FLT3" & str_detect(c., ">") ~ "FLT3-TKD",
    TRUE ~ gene
  ))
  %>% select(-p.)
  %>% left_join(variant_data, by=c("ndosbm", "c." = "nucleotidic_position"))
  %>% group_by(gene, `c.`)
  %>% fill(chromosome, position, reference, alternative,
       .direction = "downup")
  %>% ungroup()
  %>% filter(posmye != "Analyse négative")
) -> ndosbm_missing_row

(
  SQXT_ndosbm_row
  %>% bind_rows(ndosbm_missing_row)
  %>% left_join(bm_BIG %>% select(ndosbm = n_bm, ddp))
  %>% filter(!str_detect(nom, "PLAINE"))
  %>% rename(ndosbm_diag = ndosbm)
) -> all_by_row

missing_incl <- setdiff(n_incl_SGE$n_inclusion, unique(all_by_row$id))

(
  bm_BIG
  %>% filter(n_inclusion %in% missing_incl)
  %>% filter(stade == "diag")
  %>% filter(type == "MO" | type == "mO")
) -> missing_incl_diag

# path_ndosbm_missing %>% select(path) %>% write_tsv("./output/bm/path_ndosbm_missing.tsv")

(
  read_tsv("./input/bm/XML/selected_nincl_101125/missing_incl_diag.tsv")
  %>% filter(str_detect(PATH, "xml"))
) -> n_incl_path

# all_variants_data <- process_xml_folder("./input/bm/XML/selected_nincl_101125/")

(
  all_variants_data
  %>% mutate(ndosbm = str_remove(xml_file_name, "D[0-9]{2}.xml"))
  %>% select(ndosbm, nucleotidic_position, chromosome, position, reference, alternative)
  %>% distinct()
  %>% ungroup()
) -> variant_data

(
  n_incl_SGE
  %>% select(n_inclusion, ndosbm_suivi = n_bm)
  %>% left_join(POSMYE_101125 %>% select(n_inclusion = id, ndosbm_diag = ndosbm))
  %>% mutate(to_drop = FALSE, .before = "n_inclusion")
  %>% mutate(to_drop = ifelse(!str_detect(ndosbm_diag, "L"), TRUE, to_drop))
  %>% mutate(to_drop = ifelse(n_inclusion %in% c("19A-004", "22A-048", "14A-003", "01A-029"), TRUE, to_drop))
) -> correspondance_diag_suivi

(
  correspondance_diag_suivi
  %>% group_by(n_inclusion)
  %>% mutate(n = n())
  %>% ungroup()
  %>% filter(to_drop == FALSE)
  %>% left_join(all_by_row)
  %>% mutate(transcript = "")
  %>% mutate(exon = "")
  %>% left_join(bm_BIG %>% select(ndosbm_suivi = n_bm, date_suivi = ddp))
  %>% select(ndosbm_diag, lib=x7,	gene, chromosome, position, reference, 
  alternative, transcript, exon, nucleotidic_position = `c.`,	
  proteinic_position = mutation, vaf_percent = vaf, ndosbm_suivi, date_suivi,
   nom_patient = nom, DDG = ddp)
   %>% arrange(nom_patient, ndosbm_suivi)
   %>% mutate(proteinic_position = str_extract(proteinic_position, "p.*"))
) -> FC_101125

length(unique(FC_101125$ndosbm_suivi))

# writexl::write_xlsx(FC_101125, "./output/bm/FC_MAGNET_AML_101125.xlsx")

# (
#   read_xlsx("./output/bm/FC_MAGNET_AML_101125.xlsx")
#   %>% distinct(ndosbm_diag)
#   %>% nrow()
# )

```

* 81 total list SGE

* DISPO CASH 54
  * DECOURCELLE erreur meme L1706711 pour pré-greffe et J60 post allo
    * Modif 12A-010 J60 post allo L1710153
  * MISSING : "L1714853" "L1816858"
  * Modification DUPIN L1714848 to L1714853
  * ANALYSE NEGATINVE : 
    * 19A-004 LEMAIRE,Maxime
    * 22A-048 DRUJONT,Nicolas

* NON DISPO CASH 19
  * 01A-029 GILLARD BOTTEQUIN,LEA NEG 
  * A priori RUIZ SERRANO,GERARD pas de diag 
  * PLAINE,EMILIE ADN326-94_DG ?? 

* 71 total , 3 AN 1 NR = 75

* no incl "01A-029" récupérer 
* AUCUNE TRACE sur tout onco hémato "14A-003" "19A-004" "22A-048"
* même patient plusieurs samples

* Analyse négative 

* AU TOTAL 81 fichiers SGE 
  * 5 Dégager : 
    * 01A-029 NEG
    * 14A-003 NR
    * 19A-004 NEG
    * 22A-046 ADN326-94_DG ? 
    * 22A-048 NEG

* AJOUTER MANUELLEMENT c.330G>T PLAINE EMILIE

* Suppresion du variant RUNX1 NOEL,FRE
* Supression du variant FLT3 DUPIN

# 191125

* 36 p MAGNET_AML

```{r}

################################################################################
# BIG INFOS SAMPLE
################################################################################



(
  read_xlsx("./input/RUN_191125.xlsx")
  %>% clean_names()
  %>% filter(projet == "MAGNET_AML")
  # %>% mutate(
  #   ddp = case_when(
  #     str_detect(ddp, "^\\d+$") ~ as_date(as.numeric(ddp), origin = "1899-12-30"),
  #     TRUE                        ~ dmy(ddp)
  #   )
  # )
  %>% mutate(ndosbm = str_remove(prelevement, "D[0-9]{1,2}"))
  %>% select(patient, ndosbm)
  %>% mutate(to_keep = TRUE)
  # %>% select(ndosbm)
  # %>% write_tsv(., pipe("pbcopy"))
) -> list_run_mrd191125

# 36 samples 

(
  bm_BIG
  %>% left_join(list_run_mrd191125, by = c("n_bm" = "ndosbm"))
  %>% group_by(n_inclusion)
  %>% filter(any(to_keep == TRUE))
  %>% filter(to_keep == TRUE | stade == "diag")
  %>% arrange(n_inclusion, desc(stade), desc(ngs))
  %>% group_by(n_inclusion, stade)
  %>% slice(1)
  # %>% arrange(n_inclusion, desc(stade), desc(ngs))
  # %>% group_by(n_inclusion)
  # %>% mutate(n = n())
  %>% ungroup()
  %>% filter(stade == "diag")
  # %>% select(n_bm)
  # %>% write_tsv(., pipe("pbcopy"))
) -> joined

# 4 patients avec 2 points
# 32 patients donc 

# 22A-022 pas de ndosbm

################################################################################
# POSMYE
################################################################################

(
  read_xlsx("./input/bm/BASE BIG1 FINALE 2023 patients 22_07_2024_IDENTITES_POSMYE.xlsx", sheet = 2)
  %>% clean_names()
) -> raw_POSMYE

(
  bm_BIG
  %>% left_join(list_run_mrd191125, by = c("n_bm" = "ndosbm"))
  %>% filter(to_keep == TRUE)
) -> n_incl_SGE_191125

(
  raw_POSMYE
  %>% filter(id %in% n_incl_SGE_191125$n_inclusion)
) -> POSMYE_191125

(
  POSMYE_191125
  %>% filter(!is.na(x7))
  %>% mutate(x6 = str_extract(x6, "L.*"))
  # %>% select(ndosbm)
  # %>% write_tsv(., pipe("pbcopy"))
) -> SQXT_ndosbm_191125

# 3 NA 
# AUMONT, ERIC
# ARICAN, JACQUES
# JARNIGON, AXEL

# 29 patients

################################################################################
# OPEN XML
################################################################################

all_variants_data <- process_xml_folder("/Users/augustin_boudry/R/projets/251104_MAGNET_AML/input/bm/XML/191125_selected")

(
  all_variants_data
  %>% mutate(ndosbm = str_remove(xml_file_name, "D[0-9]{2}.xml"))
  %>% select(ndosbm, nucleotidic_position, chromosome, position, reference, alternative)
  %>% distinct()
  %>% ungroup()
) -> variant_data

length(unique(variant_data$ndosbm))


(
  SQXT_ndosbm_191125
  %>% separate_rows(posmye, sep = "\r\n")
  %>% mutate(gene = str_extract(posmye, list_gene_LOVER_2023))
  %>% mutate(vaf = str_extract(posmye, "\\d+(\\.\\d+)?%")) 
  %>% mutate(vaf = as.numeric(str_remove(vaf, "%")))
  %>% mutate(c. = str_extract(posmye, "c.* :"))
  %>% mutate(c. = ifelse(is.na(c.), str_extract(posmye, "c.* \\("),c.))
  %>% mutate(c. = str_remove(c., "( :)| \\("))
  %>% mutate(p. = str_extract(posmye, "p\\..*\\("))
  %>% mutate(p. = str_remove(p., "\\("))
  %>% mutate(p. = ifelse(is.na(p.), "p.intron", p.))
  %>% mutate(mutation = paste0(gene, "_", p.))
  %>% mutate(gene = case_when(
    gene == "FLT3" & str_detect(c., "dup|in") ~ "FLT3-ITD",
    gene == "FLT3" & str_detect(c., ">") ~ "FLT3-TKD",
    TRUE ~ gene
  ))
  %>% select(-p.)
  %>% left_join(variant_data, by=c("ndosbm", "c." = "nucleotidic_position"))
  %>% group_by(gene, `c.`)
  %>% fill(chromosome, position, reference, alternative,
       .direction = "downup")
  %>% ungroup()
  %>% filter(posmye != "Analyse négative")
  %>% ungroup()
) -> SQXT_ndosbm_row

length(unique(SQXT_ndosbm_row$ndosbm))

(
  POSMYE_101125
  %>% filter(is.na(x7))
  %>% mutate(x6 = str_extract(x6, "L.*"))
  # %>% select(ndosbm)
  # %>% write_tsv(., pipe("pbcopy"))
  %>% filter(posmye != "Analyse négative.")
  %>% filter(posmye != ".")
) -> ndosbm_missing

(
  n_incl_SGE_191125
  %>% select(n_inclusion, ndosbm_suivi = n_bm)
  %>% left_join(POSMYE_191125 %>% select(n_inclusion = id, ndosbm_diag = ndosbm))
  %>% mutate(to_drop = FALSE, .before = "n_inclusion")
  %>% mutate(to_drop = ifelse(!str_detect(ndosbm_diag, "L"), TRUE, to_drop))
) -> correspondance_diag_suivi


(
  correspondance_diag_suivi
  %>% group_by(n_inclusion)
  %>% mutate(n = n())
  %>% ungroup()
  %>% filter(to_drop == FALSE)
  %>% left_join(SQXT_ndosbm_row %>% rename(ndosbm_diag = ndosbm))
  %>% mutate(transcript = "")
  %>% mutate(exon = "")
  %>% left_join(bm_BIG %>% select(ndosbm_diag = n_bm, DDG = ddp))
  %>% left_join(bm_BIG %>% select(ndosbm_suivi = n_bm, date_suivi = ddp))
  %>% select(ndosbm_diag, lib=x7,	gene, chromosome, position, reference, 
  alternative, transcript, exon, nucleotidic_position = `c.`,	
  proteinic_position = mutation, vaf_percent = vaf, ndosbm_suivi, date_suivi,
   nom_patient = nom, DDG)
   #  DDG = ddp
   %>% arrange(nom_patient, ndosbm_suivi)
   %>% mutate(proteinic_position = str_extract(proteinic_position, "p.*"))
   %>% filter(ndosbm_diag != "L1602636")
   %>% filter(ndosbm_diag != "L1606609")
) -> FC_191125

writexl::write_xlsx(FC_191125, "./output/191125_FC_MAGNET_AML.xlsx")

length(unique(FC_191125$nom_patient))


```

# TW-MRD-021225
* 81 - 1 Ø de diag
```{r}
(
  read_xls("./input/bm/TW-MRD-031225-DRAGEN.xls")
  %>% clean_names()
  %>% filter(projet == "MAGNET_AML")
  %>% count(patient)
  %>% filter(n > 1)
) -> nom_run

# 6 patients en doubles 

(
  read_xls("./input/bm/TW-MRD-101125-DRAGEN.xls")
  %>% clean_names()
  %>% filter(projet == "MAGNET_AML")
  %>% select(prelevement)
  %>% mutate(to_keep = TRUE)
  %>% mutate(n_bm = str_remove(prelevement, "D[0-9]{2}"))
) -> ndosbm_run

(
  bm_BIG
  %>% left_join(ndosbm_run)
  %>% group_by(n_inclusion)
  %>% filter(any(to_keep == TRUE))
  %>% filter(to_keep == TRUE | stade == "diag")
  %>% arrange(n_inclusion, desc(stade), desc(ngs))
  %>% group_by(n_inclusion, stade)
  %>% slice(1)
  %>% arrange(n_inclusion, desc(stade), desc(ngs))
  %>% ungroup()
  %>% group_by(n_inclusion)
  %>% mutate(n = n())
) -> joined

(
  joined
  %>% select(n_inclusion, stade, n_bm, ngs)
  # %>% count(n_inclusion)
) -> selected

# 81 - 6 = 75 OK on a tout les diag !!! 

(
  selected
  %>% filter(stade == "diag")
) -> filtered_diag

# 22A-046 PAS DE DIAG

# (
#   selected
#   %>% mutate(stade = ifelse(stade != "diag", "suivi", stade))
#   %>% pivot_wider(names_from = stade, values_from = n_bm)
# ) -> cleaned

(
  raw_POSMYE
  %>% filter(id %in% filtered_diag$n_inclusion)
  %>% filter(!str_detect(posmye, "négative"))
  %>% filter(posmye != ".")
) -> POSMYE_021225

# - 3 NEG - 1 "." = 71 - 1 KD6MA 05A-056

(
  POSMYE_021225
  %>% filter(!is.na(x6))
  %>% mutate(x6 = str_extract(x6, "L.*"))
) -> SSQXT_POSMYE_021225

(
  SSQXT_POSMYE_021225
  %>% select(x6)
  %>% write_tsv(., pipe("pbcopy"))
)

all_variants_data <- process_xml_folder("./input/bm/XML/selected_021225/")

(
  all_variants_data
  %>% mutate(ndosbm = str_remove(xml_file_name, "D[0-9]{2}.xml"))
  %>% select(ndosbm, nucleotidic_position, chromosome, position, reference, alternative)
  %>% distinct()
  %>% ungroup()
) -> variant_data

length(unique(variant_data$ndosbm))

(
  SSQXT_POSMYE_021225
  %>% separate_rows(posmye, sep = "\r\n")
  %>% mutate(gene = str_extract(posmye, list_gene_LOVER_2023))
  %>% mutate(vaf = str_extract(posmye, "\\d+(\\.\\d+)?%")) 
  %>% mutate(vaf = as.numeric(str_remove(vaf, "%")))
  %>% mutate(c. = str_extract(posmye, "c.* :"))
  %>% mutate(c. = ifelse(is.na(c.), str_extract(posmye, "c.* \\("),c.))
  %>% mutate(c. = str_remove(c., "( :)| \\("))
  %>% mutate(p. = str_extract(posmye, "p\\..*\\("))
  %>% mutate(p. = str_remove(p., "\\("))
  %>% mutate(p. = ifelse(is.na(p.), "p.intron", p.))
  %>% mutate(mutation = paste0(gene, "_", p.))
  %>% mutate(gene = case_when(
    gene == "FLT3" & str_detect(c., "dup|in") ~ "FLT3-ITD",
    gene == "FLT3" & str_detect(c., ">") ~ "FLT3-TKD",
    TRUE ~ gene
  ))
  %>% select(-p.)
  %>% left_join(variant_data, by=c("ndosbm", "c." = "nucleotidic_position"))
  %>% group_by(gene, `c.`)
  %>% fill(chromosome, position, reference, alternative,
       .direction = "downup")
  %>% ungroup()
  %>% filter(posmye != "Analyse négative")
  %>% ungroup()
  %>% mutate(chromosome = ifelse(nom == "DIJON,VALERIE" & gene == "NPM1", "chr5", chromosome))
  %>% mutate(position = ifelse(nom == "DIJON,VALERIE" & gene == "NPM1", 170837544, position))
  %>% mutate(alternative = ifelse(nom == "DIJON,VALERIE" & gene == "NPM1", "TCTGC", alternative))
  %>% mutate(reference = ifelse(nom == "DIJON,VALERIE" & gene == "NPM1", "T", reference))
  %>% filter(gene != "KDM6A")
) -> SQXT_ndosbm_row

length(unique(SQXT_ndosbm_row$ndosbm))

# (
#   POSMYE_101125
#   %>% filter(is.na(x7))
#   %>% mutate(x6 = str_extract(x6, "L.*"))
#   # %>% select(ndosbm)
#   # %>% write_tsv(., pipe("pbcopy"))
#   %>% filter(posmye != "Analyse négative.")
#   %>% filter(posmye != ".")
# ) -> ndosbm_missing

(
  selected
  %>% select(-ngs)
  %>% mutate(stade = ifelse(stade != "diag" | is.na(stade), "suivi", stade))
  %>% pivot_wider(names_from = "stade", values_from = n_bm)
  %>% mutate(
    suivi = purrr::map_chr(suivi, ~ paste(.x, collapse = ","))
  )
  %>% mutate(
    diag = purrr::map_chr(diag, ~ paste(.x, collapse = ","))
  )  
  %>% separate_rows(suivi, sep = ",")
  %>% rename(ndosbm_diag = diag)
  %>% rename(ndosbm_suivi = suivi)
  # n_incl_SGE_191125
  # %>% select(n_inclusion, ndosbm_suivi = n_bm)
  # %>% left_join(POSMYE_191125 %>% select(n_inclusion = id, ndosbm_diag = ndosbm))
  # %>% mutate(to_drop = FALSE, .before = "n_inclusion")
  # %>% mutate(to_drop = ifelse(!str_detect(ndosbm_diag, "L"), TRUE, to_drop))
) -> correspondance_diag_suivi


(
  correspondance_diag_suivi
  %>% group_by(n_inclusion)
  %>% mutate(n = n())
  %>% ungroup()
  # %>% filter(to_drop == FALSE)
  %>% left_join(SQXT_ndosbm_row %>% rename(ndosbm_diag = ndosbm))
  %>% mutate(transcript = "")
  %>% mutate(exon = "")
  %>% left_join(bm_BIG %>% select(ndosbm_diag = n_bm, DDG = ddp))
  %>% left_join(bm_BIG %>% select(ndosbm_suivi = n_bm, date_suivi = ddp))
  %>% select(ndosbm_diag, lib=x7,	gene, chromosome, position, reference, 
  alternative, transcript, exon, nucleotidic_position = `c.`,	
  proteinic_position = mutation, vaf_percent = vaf, ndosbm_suivi, date_suivi,
   nom_patient = nom, DDG)
   #  DDG = ddp
   %>% arrange(nom_patient, ndosbm_suivi)
   %>% mutate(proteinic_position = str_extract(proteinic_position, "p.*"))
   %>% filter(!is.na(gene))
) -> FC_021225

# HERE 

writexl::write_xlsx(FC_021225, "./output/021225_FC_MAGNET_AML.xlsx")

length(unique(FC_021225$ndosbm_suivi))


```